packages/flutter_tools/lib/src/web/devfs_web.dart

```diff
@@ -256,18 +247,110 @@ class WebAssetServer {
+    File file = globals.fs.file(globals.fs.currentDirectory.uri.resolve(path));
+    final List<String> segments = path.split('/');
+    if (segments.first.isEmpty) {
+      segments.removeAt(0);
+    }
+    // If both of the lookups above failed, the file might have been a package
+    // file which is signaled by a `/packages/<package>/<path>` request.
+    if (!file.existsSync() && segments.first == 'packages') {
+      file = globals.fs.file(_packages.resolve(Uri(
+          scheme: 'package', pathSegments: segments.skip(1))));
+    }
+    // If it isn't a project source or an asset, it must be a dart SDK source.
+    // or a flutter web SDK source.
+    if (!file.existsSync()) {
+      final Directory dartSdkParent = globals.fs
+          .directory(
+              globals.artifacts.getArtifactPath(Artifact.engineDartSdkPath))
+          .parent;
+      file = globals.fs.file(globals.fs.path
+          .joinAll(<String>[dartSdkParent.path, ...segments]));
+    }
+
+    if (!file.existsSync()) {
+      final String flutterWebSdk =
+          globals.artifacts.getArtifactPath(Artifact.flutterWebSdk);
+      file = globals.fs.file(globals.fs.path
+          .joinAll(<String>[flutterWebSdk, ...segments]));
+    }
+
+    return file;
+  }
+
+  @override
+  Future<String> dartSourceContents(String serverPath) {
+    final File result = _resolveDartFile(serverPath);
+    if (result.existsSync()) {
+      return result.readAsString();
+    }
+    return null;
+  }
+
+  @override
+  Future<String> sourceMapContents(String serverPath) async {
+    return utf8.decode(_sourcemaps[serverPath]);
+  }
+}
+
+class ConnectionResult {
+  ConnectionResult(this.appConnection, this.debugConnection);
+
+  final AppConnection appConnection;
+  final DebugConnection debugConnection;
 }
 
 class WebDevFS implements DevFS {
-  WebDevFS(this.hostname, this.port, this._packagesFilePath);
+  WebDevFS({
+    @required this.hostname,
+    @required this.port,
+    @required this.packagesFilePath,
+    @required this.urlTunneller,
+    @required this.buildMode,
+    @required this.enableDwds,
+  });
 
   final String hostname;
   final int port;
-  final String _packagesFilePath;
+  final String packagesFilePath;
+  final UrlTunneller urlTunneller;
+  final BuildMode buildMode;
+  final bool enableDwds;
 
   @visibleForTesting
   WebAssetServer webAssetServer;
 
+  Dwds get dwds => webAssetServer.dwds;
+
+  Future<DebugConnection> _cachedExtensionFuture;
+  StreamSubscription<void> _connectedApps;
+
+  /// Connect and retrieve the [DebugConnection] for the current application.
+  ///
+  /// Only calls [AppConnection.runMain] on the subsequent connections.
+  Future<ConnectionResult> connect(bool useDebugExtension) {
+    final Completer<ConnectionResult> firstConnection = Completer<ConnectionResult>();
+    _connectedApps = dwds.connectedApps.listen((AppConnection appConnection) async {
+      final DebugConnection debugConnection = useDebugExtension
```

Comment:
If this `listen` callback throws an exception it will be an unhandled exception.

Reply:
Updated to forward the error through the completer

Reply:
`onError` handles errors on the stream, but won't run if the listen callback throws.